### PR TITLE
Store sensor info in boost::ptree and output to JSON

### PR DESCRIPTION
**Pull request by rradjabi**
This is a work in progress and it needs some feedback.

### CODE REVIEW

**Comment by rradjabi** [comment]
Refs #632

File: src/runtime_src/driver/xclng/tools/xbutil/xbutil.h [hunk 1]
```diff
@@ -207,32 +210,53 @@ class device {
         return 0;
     }
 
+    int parseComputeUnits(std::vector<ip_data> &computeUnits) const
```
Comment (sonals) [inline src/runtime_src/driver/xclng/tools/xbutil/xbutil.h]: **const** std::vector<ip_data> &computeUnits)

File: src/runtime_src/driver/xclng/xrt/user_common/sensor.cpp [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+//    root.put( "board.firewall", "" );
+//    root.put( "board.xclbin", "" );
+//    root.put( "board.memory", "" );
+//    root.put( "board.stream", "" );
+//    root.put( "board.compute_unit", "" );
+    
+    return 0;
+}
+
+int writeTree( boost::property_tree::ptree &root )
```
Comment (sonals) [inline src/runtime_src/driver/xclng/xrt/user_common/sensor.cpp]: **const** boost::property_tree::ptree &root

File: src/runtime_src/driver/xclng/tools/xbutil/xbutil.h [hunk 3]
```diff
@@ -207,32 +211,53 @@ class device {
...
```
Comment (maxzhen) [inline src/runtime_src/driver/xclng/tools/xbutil/xbutil.h]: Maybe we should stop adding implementation code in header file...

File: src/runtime_src/driver/xclng/tools/xbutil/xbutil.h [hunk 4]
```diff
@@ -722,6 +775,239 @@ class device {
 
         lines.push_back(ss.str());
     }
+    
+    int readSensors( void ) const
```
Comment (maxzhen) [inline src/runtime_src/driver/xclng/tools/xbutil/xbutil.h]: We are not reading just sensor output here. Maybe we should split the function into two: one is for reading static info and the other is for dynamic info. We may need to call dynamic one in xbutil top repeatedly to get up-to-date info.

**Comment by rradjabi** [comment]
retest this please

File: src/runtime_src/driver/xclng/tools/xbutil/xbutil.h (hunk 3 above)
Comment (rradjabi) [reply src/runtime_src/driver/xclng/tools/xbutil/xbutil.h]: I agree, I will try to move this into the sensor files but that's beyond the scope of this feature.

**Comment by nbalagopal** [comment]
retest this please

File: src/runtime_src/driver/xclng/xrt/user_common/sensor.h [hunk 5]
```diff
@@ -0,0 +1,12 @@
+#ifndef SENSOR_H
+#define SENSOR_H
+
+#include <boost/property_tree/ptree.hpp>
+#include <boost/property_tree/json_parser.hpp>
+#include <boost/foreach.hpp>
+
+static boost::property_tree::ptree gSensorTree;
```
Comment (stsoe) [inline src/runtime_src/driver/xclng/xrt/user_common/sensor.h]: No static variables in header files.  If the header is include multiple times, each compilation unit will have an instance of the static variable.  Move to sensor.cpp

Comment (rradjabi) [reply src/runtime_src/driver/xclng/xrt/user_common/sensor.h]: @stsoe 
Please review my latest changes. I made this a singleton class where the ptree object is a private member variable. I've written wrappers around ptree::get and ptree::put.

Comment (stsoe) [reply src/runtime_src/driver/xclng/xrt/user_common/sensor.h]: I am confused, where is the global used?  Who calls Instance()?

Comment (rradjabi) [reply src/runtime_src/driver/xclng/xrt/user_common/sensor.h]: The property tree (previously gSensorTree) is no-longer global. Instance() provides a reference to the singleton class where sensorTree::put() and sensorTree::get() are wrappers around boost::property_tree::ptree::put() and ::get().

Comment (stsoe) [reply src/runtime_src/driver/xclng/xrt/user_common/sensor.h]: I still don't see where Instance() is called?

Comment (rradjabi) [reply src/runtime_src/driver/xclng/xrt/user_common/sensor.h]: In xbutil.h (See line 239) wherever sensorTree::put() is called it must be called like so:


```
sensorTree &tree = sensorTree::Instance();
tree.put( val );
```

Comment (stsoe) [reply src/runtime_src/driver/xclng/xrt/user_common/sensor.h]: ok. thanks.

**Pull request merged**